projects/web-wallet/rpc-protocol.md

```diff
@@ -0,0 +1,432 @@
+```
+
+## 2.1.8 Error codes
+
+We should follow the [JSON-RPC](https://www.jsonrpc.org/specification#error_object) specification for error codes, so errors generated by the "server", which in our case, is the wallet handling the requests should be in the `-32099` to `-32000` range.
+
+| Error | Error Type | Description                        |
+|------------|------------------------------------|---|
+| -32050     | WALLET_NOT_READY | Wallet is not ready yet              |
+| -32051     | REQUEST_REJECTED | Request rejected by the user |
+| -32052     | MISSING_FUNDS | Missing funds to fulfill transaction |
+| -32053 | INVALID_PARAMETERS | Invalid parameters |
+| -32099     | GENERIC_FAILURE | Generic failure |
+
+# 3. Guide-level Explanation
+
+### 3.1 `htr_sendTx`
+
+When the request is received, we will validate if the parameters are valid, enforcing the data types and rules described in the Reference-Level explanation
+
+If the validation fails, we will respond with a `INVALID_PARAMETERS` error code
+
+If the validation is successful, we will display a confirmation screen, with all outputs for manual user validation (UX is yet to be discussed). If the user rejects the transaction, we will respond with a `REQUEST_REJECTED` error
+
+If the user accepts the transaction, we will create a `SendTransaction` model (or `SendTransactionWalletService` if on the wallet-service) and execute the send transaction.
+
+Since `JSON-RPC` is designed considering only a single response per request, we will only respond the request when the transaction either failed to be pushed to the fullnode (or the wallet-service) or when it's successfully pushed.
+
+## 3.2 `htr_createToken`
+
+When the request is received, we will validate if the parameters are valid by following the parameter type definitions in the Reference-Level explanation
+
+If the validation fails, we will respond with a `INVALID_PARAMETERS` error code
+
+At this point, we will call `prepareCreateNewToken` and display the assembled transaction for user validation (UX is yet to be discussed).
+
+If the user accepts the transaction, we will, just like the `htr_sendTx` method, create a `SendTransaction` model and execute the send transaction. Otherwise, we will answer with a `REQUEST_REJECTED` error
+
+## 3.3 `htr_getUtxos`
+
+When the request is received, we will validate if the parameters are valid by following the parameter type definitions in the Reference-Level explanation
+
+If the validation fails, we will respond with a `INVALID_PARAMETERS` error code
+
+At this point, we will call the `getUtxos` method to retrieve `utxos` that fulfill the requested filters.
+
+Before returning, we will display a confirmation screen (UX yet to be discussed) asking for user confirmation by displaying the `utxos` that will be returned to the `dApp`.
+
+If the user accepts the transaction, we will reply with the array of `utxos`. Otherwise, we will answer with a `REQUEST_REJECTED` error
+
+#### Future possibilities
```

Comment:
I guess the context here is not for future possibilities. Please double check.

Reply:
What I mean here is that I'm suggesting that we don't allow the user to select which utxo he wants to return, he can only reject or approve

In the future, we might allow him to select exactly which one he approves and even replace it with another one

Reply:
Got it. Thanks!